src/System.Private.CoreLib/shared/System/Runtime/InteropServices/GCHandle.cs

```diff
@@ -0,0 +1,164 @@
+
+        /// <summary>Creates a new GC handle for an object.</summary>
+        /// <param name="value">The object that the GC handle is created for.</param>
+        /// <param name="type">The type of GC handle to create.</param>
+        /// <returns>A new GC handle that protects the object.</returns>
+        public static GCHandle Alloc(object value, GCHandleType type) => new GCHandle(value, type);
+
+        /// <summary>Frees a GC handle.</summary>
+        public void Free()
+        {
+            // Free the handle if it hasn't already been freed.
+            IntPtr handle = Interlocked.Exchange(ref m_handle, IntPtr.Zero);
+            ThrowIfInvalid(handle);
+            InternalFree(GetHandleValue(handle));
+        }
+
+        // Target property - allows getting / updating of the handle's referent.
+        public object Target
+        {
+            get
+            {
+                ThrowIfInvalid();
+                return InternalGet(GetHandleValue());
+            }
+            set
+            {
+                ThrowIfInvalid();
+                InternalSet(GetHandleValue(), value, IsPinned);
+            }
+        }
+
+
+        /// <summary>
+        /// Retrieve the address of an object in a Pinned handle.  This throws
+        /// an exception if the handle is any type other than Pinned.
+        /// </summary>
+        public IntPtr AddrOfPinnedObject()
+        {
+            // Check if the handle was not a pinned handle.
+            // You can only get the address of pinned handles.
+            ThrowIfInvalid();
+            if (!IsPinned)
+            {
+                ThrowHelper.ThrowInvalidOperationException_HandleIsNotPinned();
+            }
+
+            // Get the address.
+            return InternalAddrOfPinnedObject(GetHandleValue());
+        }
+
+        /// <summary>Determine whether this handle has been allocated or not.</summary>
+        public bool IsAllocated => m_handle != IntPtr.Zero;
+
+        /// <summary>
+        /// Used to create a GCHandle from an int.  This is intended to
+        /// be used with the reverse conversion.
+        /// </summary>
+        public static explicit operator GCHandle(IntPtr value)
+        {
+            ThrowIfInvalid(value);
+            return new GCHandle(value);
+        }
+
+        public static GCHandle FromIntPtr(IntPtr value)
+        {
+            ThrowIfInvalid(value);
+            return new GCHandle(value);
+        }
+
+        /// <summary>Used to get the internal integer representation of the handle out.</summary>
+        public static explicit operator IntPtr(GCHandle value) => ToIntPtr(value);
+
+        public static IntPtr ToIntPtr(GCHandle value) => value.m_handle;
+
+        public override int GetHashCode() => m_handle.GetHashCode();
+
+        public override bool Equals(object o) => o is GCHandle && m_handle == ((GCHandle)o).m_handle;
```

Comment:
```suggestion
        public override bool Equals(object o) => o is GCHandle gch && m_handle == gch.m_handle;
```

Reply:
I've been shying away from using that with value types because of historically worse codegen. It's supposed to be fixed with recent compiler bits, but I've not validated that. In going to leave this as is for now.